### PR TITLE
Application launch hooks

### PR DESCRIPTION
**Pull request by iLLiCiTiT**
## Changes
- Implemented applications launch hooks
    - Hooks are executed before or after application launch and has access to launching context `<ApplicationLaunchContext>`
    - They are executed by order (attribute on hook) **lowest** -> **highest** -> **`None`**
    - Pre launch hooks:
        - Are executed before application process is opened.
        - Are meant for changing environments, modify launch arguments, prepare workfile from template etc.
        - If one will crash others won't continue and application won't be started.
        - If crash is caused by known issue it is expected to raise `ApplicationLaunchFailed` exception with user readable message.
    - Post launch hooks:
        - At the moment won't affect anything and error messages are not captured and showed.
        - Crash of one won't affect others.
- New classes:
    - `LaunchHook` - base abstract class of launch hook which contain launch hook logic, should not be used as base for hooks
    - `PreLaunchHook` - inherit from `LaunchHook` currently do not contain any modifications in logic - abstract class for prelaunch hooks
    - `PostLaunchHook` - inherit from `LaunchHook` currently do not contain any modifications in logic - abstract class for postlaunch hooks
    - currently classes `PreLaunchHook` and `PostLaunchHook` are implemented only to help differentiate if should be executed before or after application launch, but may contain different code in the future
    - `LaunchHook` on the other hand should not be used as base class for hooks. It is possible to do so but hook will never load.

- Launch hooks are file based python scripts which must contain class definition which inherit from classes `PreLaunchHook` or `PostLaunchHook`
    - I used prefix for hook filenames `pre_*` and `post_*` to differentiate between them. (It is not requirement)
    - Filenames starting with underscore `_` are skipped.

- Converted (and modified) current hooks and added few new:
    - Created global pre launch hook which works only on host applications.
        - Prepare host environments, context environments and collect data about last workfile
        - Copied code which was in `ApplicationLaunchContext` and was executed all the time
    - Created global post launch hook which cares about ftrack changes (change status, start timer etc.)
        - Copied code which was in `ApplicationLaunchContext` and was executed all the time
    - Implemented prelaunch hooks for NukeX, NukeStudio, Hiero to add launch argument after executable (and add last workfile path)
    - Implemented prelaunch hooks for `Photoshop`, `AfterEffect`, `Harmony` and `TVPaint` where python script must be executed instead of application executable
        - Tested:
            - [x] Photoshop
            - [x] AfterEffect
            - [x] TVPaint
            - [ ] Harmony
    - Converted `Unreal`, `Celaction` and `Fusion` hooks
        - Tested:
            - [ ] Unreal
            - [ ] Celaction
            - [ ] Fusion

- Ftrack sync to avalon can use application manager for validations of applications on sync
- added `TVPaint`, `Adobe AfterEffects` (2020, 2021) and `Adobe Photoshop 2021` to applications settings

## Questions
- Should be `LaunchHook` renamed to `_LaunchHook` so it's visible it should not be used.
- I don't like **post**hook, are you OK with **post**? Any better suggestions?

## TODO
- Better hooks paths getter + how to add (register) different paths to hooks (for Ftrack etc.).

### CODE REVIEW

File: pype/hooks/nukestudio/pre_launch_args.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import os
+from pype.lib import PreLaunchHook
+
+
+class NukeStudioLaunchArguments(PreLaunchHook):
+    order = 0
+    hosts = ["nukestudio"]
```
Comment (jezscha) [inline pype/hooks/nukestudio/pre_launch_args.py:7]: this had been changed to `hiero` host already `nukestudio` want work anywhere as host.

Comment (jezscha) [reply pype/hooks/nukestudio/pre_launch_args.py:7]: that apply to the name of folder

Comment (jezscha) [inline pype/hooks/nukestudio/pre_launch_args.py:7]: `nukestudio` host name had been renamed to `hiero` and it apply to the name of folder too.

**Review by jezscha — CHANGES_REQUESTED** [review]

**Comment by mkolar** [comment]
With this, we are limiting the launch arguments in the executables. We should add the option to add arguments directly to the executable field.

Let's use executable field directly to verify it, in cases where it doesn't contain `""`, but if we find quotes in the path, we'll consider it's content to be executable and everything else it's arguments. that way we can still tweak arguments directly in the settings without touching the pre-hooks, and pre-hooks will only append extras

image
image

**Comment by jezscha** [comment]
> With this, we are limiting the launch arguments in the executables. We should add the option to add arguments directly to the executable field.

@mkolar wouldn't be more user friendly to add app attribute `launch_arguments`?

**Comment by iLLiCiTiT** [comment]
As we discussed on friday it would be safer to separate executable and arguments to different inputs.

Before I can do that there are needed changes from https://github.com/pypeclub/pype/pull/762/files so I have to wait for that.

How it may look:
image

File: pype/hooks/nukestudio/pre_launch_args.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class NukeStudioLaunchArguments(PreLaunchHook):
+    order = 0
+    app_groups = ["nukestudio"]
+
+    def execute(self):
+        """Prepare suprocess launch arguments for NukeStudio."""
+        # Add path to workfile to arguments
+        if self.data.get("start_last_workfile"):
```
Comment (mkolar) [inline pype/hooks/nukestudio/pre_launch_args.py:12]: Do we set this in some way currently? I assume that logic willing this hook `self.data` is still missing.

Comment (iLLiCiTiT) [reply pype/hooks/nukestudio/pre_launch_args.py:12]: That is set in global host prelaunch hook `~/hooks/global/pre_global_host_data.py` https://github.com/pypeclub/pype/pull/747/files#diff-cd623e011c6a95b4660d1c8d06aba5d85d39b8c7d84e6dd4583d11bb262c45f2R225

**Pull request merged**